### ISSUE TITLE
Schema design: User model

### ISSUE THREAD

**Issue by josueetcom** [beyondhb1079/s4us#58]
We need to establish what a User object consists of and how it will be stored in Firestore.

Things to consider:
- Firestore is a NoSQL, document-oriented database. See their [Data model](https://cloud.google.com/firestore/docs/data-model).
- See [Structuring data](https://cloud.google.com/firestore/docs/concepts/structure-data) for tips on how to structure data and [Data types](https://cloud.google.com/firestore/docs/concepts/data-types) for info on what data types are supported.

The expectation here is to come up with the User model in some form.

**Comment by josueetcom** [comment]
Some initial thoughts, a User object should probably have at least:

- an id
- personal details, e.g. age and year in school
- a role
- scholarship preferences
- other settings/preferences

**Comment by edwingl10** [comment]
Firebase Auth can contain the users:
displayName: String
email: String
emailVerified: boolean
phoneNumber: String
PhotoURL: String
uid: String

And our User model can contain other information such as: 
age: int
schoolYear: int
role: String
gender: String
scholarshipPreferences: {
type: String
major: String
amount: int
school: String
grade: int
location: String
}
other settings{
notifications: boolean
deadlineNotifications: boolean
relatedNotifications: boolean
}

**Comment by josueetcom** [comment]
> Firebase Auth can contain the users:
> displayName: String
> email: String
> emailVerified: boolean
> phoneNumber: String
> PhotoURL: String
> uid: String

Right, that's a Firebase User's [properties](https://firebase.google.com/docs/reference/js/firebase.User#displayname)

> And our User model can contain other information such as:
> age: int

Maybe birth date would be better?

> schoolYear: int

I imagine this would probably be better served as an enum, though in Firestore it might be better stored as a string than an integer.

> role: String
> gender: String

These are probably enums as well.

> scholarshipPreferences: {
> type: String

Likely an enum

> major: String

This could be an array of majors.

> amount: int

This could be a range and maybe we could support "full ride" too.

> school: String

Probably an array of schools as well if they're interested in applying to multiple schools

> grade: int

I'm curious what this would look like if they're interested in different types of schools (e.g. community college, university, vocational and technical schools, etc.)

> location: String

By location do you mean state? If so, maybe this can be an array of states in case they're looking at schools in various places?
 
> other settings: {
> notifications: boolean
> deadlineNotifications: boolean
> relatedNotifications: boolean
> }

Hmm, I would maybe call this "notification_settings" and then use enums here instead of boolean in case we support different notification settings, e.g. emails yes but mobile no, mobile yes but email no, etc.

Overall, I generally like what we're storing here, I think there's some design considerations to be made though. In general, I suggest designing for worst case scenarios for each of these fields and asking questions about what does and doesn't make sense. E.g. Does it make sense for this field be repeatable (an array)? Can this field be empty? For truthy fields, could this potentially become more than two options? For string fields, is this more like an arbitrary string or a value from a fixed set of possibilities (enum)? Can we save the user from having to update this field (in the case of age vs. birth date)?

**Comment by josueetcom** [comment]
Other questions that come to my mind are
- How precise do we need to be? E.g. For birth date do we really need the day as well or can we sacrifice precision for storing less sensitive birth info, e.g. birth month + year or just birth year. Obviously storing birth year is least sensitive user data but perhaps that means when we try recommending/filtering info based on age we might give off by one year of age results. With birth + month it's at most off by one month. Though perhaps we can let the user choose how much of that data they want to give
- How will this data be used in queries? E.g. it's easier to filter by a string field than to filter by array fields that contain a string. Perhaps something would be better stored as an array but, if a single value field covers 95% of uses cases then for performance reasons that might be better. Or alternatively a field could have a secondary values array field that stores the other values but keeps a primary one stored separately for performance reasons.

Lots of things to consider

**Comment by edwingl10** [comment]
> > grade: int
> 
> I'm curious what this would look like if they're interested in different types of schools (e.g. community college, university, vocational and technical schools, etc.)

for now I made it an enum with values like "freshman", "sophmore" etc. instead of number values.



>     With birth + month it's at most off by one month. Though perhaps we can let the user choose how much of that data they want to give
I agree. If the user chooses to not provide a DOB, we wont recommend any scholarships based on age. If they just give a year, we can calculate a range between 2 ages and recommend based on those years. And if they give a month + year then it'll help us be more precise.

**Comment by edwingl10** [comment]
Updated User model:

DOB: date or undefined
school Year: enum (unknown, highschool junior, college freshman etc)
Role: enum(Reader, Contributor, Moderator, Admin)
gender: enum(unknown, male, female, other)
scholarship Preferences: {
    type: enum[](unknown, academic, sports, community, organization)
    major: string[] or undefined
    amount: {
    type: enum[](unknown, fixed, range, full)
    min: number or undefined
    max: number or undefined
    }
school string[]
school year: (unknown, highschool junior, college freshman etc)
state: string[]
}
Notification Settings: {
    notifications: enum(none, email, mobile)
    deadline notifications: enum(same as above)
    related notifications: enum(same as above)
}

**Comment by josueetcom** [comment]
> school year: (unknown, highschool junior, college freshman etc)

Hmm, shouldn't this just be their current year?

> notifications: enum(none, email, mobile)

"general notifications" maybe

Some other personal info we could collect are: school, state, year, major(s). This could help distinguish from preferences when we're trying to recommend scholarships for students applying to schools vs students already in college.

**Comment by edwingl10** [comment]
> Hmm, shouldn't this just be their current year?

yes, that is what I meant, I'll change the name to avoid confusion

> Some other personal info we could collect are: school, state, year, major(s)

sounds good, I'll update it.
